qermit/spectral_filtering/spectral_filtering.py

```diff
@@ -0,0 +1,642 @@
+    )
+
+
+def gen_obs_exp_grid_gen_task() -> MitTask:
+    """Generates task creating an ObservableExperiment for each point in the 
+    inputted meshgrid.
+
+    :return: Task generating grid of experiments.
+    :rtype: MitTask
+    """
+    
+    def task(
+        obj,
+        obs_exp_list:List[ObservableExperiment],
+        obs_exp_sym_val_grid_list:List[NDArray]
+    ) -> Tuple[List[NDArray]]:
+        """Task generating a grid of ObservableExperiments.
+        Each point in the grid corresponds to substituting the symbols in
+        the circuit of each `ObservableExperiment` for a value in the
+        meshgrids of `obs_exp_sym_val_grid_list`.
+
+        :param obs_exp_list: A list of `ObservableExperiments`.
+        :type obs_exp_list: list[ObservableExperiment]
+        :param obs_exp_sym_val_grid_list: A list of collections of meshgrids.
+            Each symbol in the Circuit of each `ObservableExperiment` is
+            represented in the meshgrids of `obs_exp_sym_val_grid_list`.
+        :type obs_exp_sym_val_grid_list: list[NDArray[float]]
+        :return: A grid of `ObservableExperiments`. Each `ObservableExperiment`
+            includes a `SymbolDict` evaluated at a point in the grid of
+            `obs_exp_sym_val_grid_list`.
+        :rtype: tuple[list[NDArray[ObservableExperiment]]]
+        """
+        
+        # A grid of ObservableExperiment is generated for each
+        # ObservableExperiment in obs_exp_list
+        obs_exp_grid_list = []
+        for obs_exp, sym_val_grid_list in zip(obs_exp_list, obs_exp_sym_val_grid_list):
+                      
+            # List of symbols in circuit
+            sym_list = list(obs_exp.AnsatzCircuit.SymbolsDict.symbols_list)
+            # Initialise empty grid of ObservableExperiment
+            obs_exp_grid = np.empty(
+                sym_val_grid_list[0].shape,
+                dtype=ObservableExperiment
+            )
+                                    
+            # Generate an ObservableExperiment for every symbol
+            # value in the grid
+            grid_point_val_list = [
+                [i for i in range(size)]
+                for size in sym_val_grid_list[0].shape
+            ]
+            for grid_point in product(*grid_point_val_list):
+                                                
+                # Generate dictionary mapping every symbol to it's value at
+                # the given point in the grid.
+                sym_map = {
+                    sym:sym_val_grid[grid_point]
+                    for sym_val_grid, sym in zip(sym_val_grid_list, sym_list)
+                }  
+                sym_dict = SymbolsDict().symbols_from_dict(sym_map)
+                circ = obs_exp.AnsatzCircuit.Circuit.copy()
+                anz_circ = AnsatzCircuit(
+                    circ,
+                    obs_exp.AnsatzCircuit.Shots,
+                    sym_dict
+                )
+                # This needs to be a deep copy
```

Comment:
Can you explain why it needs to be a deep copy please

Reply:
I seem not to be able to reproduce the error which occurred without the deep copy before. This makes me think that this was a separate problem. Would you be surprised if a deep copy was required?